_drafts/debug-vs-release-builds.md

```diff
@@ -0,0 +1,202 @@
+information, such as data structures, variables, and the call stack, is present.
+I've never hit a roadblock due to compiler optimization in step-debugging. In the worst
+case, I need to peek at a few assembly instructions to see what's going on, but
+that's rare.
+
+There's a HUGE downside to having different optimization levels for debug vs.
+release builds: we end up with a **different** executable:
+
+- sizes will likely be different: stack or heap usage (affecting program
+  stability!), code space (affecting flash memory allocation, OTA)
+- different machine code executed by the CPU: this can impact **timing** (often
+  where the nastiest bugs show up) and actual **program validity**, e.g.
+  assumptions you are making about certain spots in your program may no longer
+  be valid with different optimization levels (the classic one is instruction
+  reordering causing concurrent access bugs!)
+
+Because your debug (testing) build is not running the same code, you may see
+these bugs only show up during field test, rather than when the device is on
+your desk and connected to a debugger!
+
+It's also generally not desirable to adjust compiler optimization level when a
+project is nearing time to ship, to decrease risk. If the debug + release builds
+are using the same optimization settings, it's less likely to hit a surprise
+out-of-space issue that can stop a project in its tracks 😱.
+
+**Summary:**
+
+- use the same optimization settings for all builds to decrease risk
+
+## `NDEBUG` preprocessor definition
+
+Often these Debug/Release builds will have some preprocessor defined symbols
+such as:
+
+- `DEBUG=1` (debug enabled)
+- `NDEBUG=0` (indicating this is NOT a non-debug build)
+
+Usually these flags will be used to compile in features like this to the Debug
+builds:
+
+- enable an external CLI for testing (this is so useful in production devices! just
+  secure it properly)
+- disable runtime asserts (assuming your system can tolerate a reset due to an
+  `ASSERT` condition, these are stupendously valuable in production builds- see
+  [this post]({% post_url 2019-11-20-cortex-m-hardfault-debug %}) for more
+  details)
+- changing default log levels, i.e. enabling `INFO` or `DEBUG` levels- this can
+  impact timing or memory/bandwidth usage, so generally it's preferable to stick
+  to one level and only increase in select cases when you need extra information
+  (eg enable it only for a single file actively being debugged)
+
+I've seen the `DEBUG` flag used pretty aggressively to cut out entire subsystems
+that aren't strictly required for production. This can cause a lot of confusion,
+for example the following scenario:
+
+1. production test build encounters a problem
+2. engineer goes to retrieve logs, realizes the logging subsystem was entirely
+   disabled in the "release" build
+3. 😩
+
+**Summary:**
+
+- it can be quite hazardous to allow splitting features based on preprocessor
```

Comment:
so you are recommending keeping the features in and then doing configuration (feature flags) over preprocessor usage?